### PR TITLE
[Feature] Add dimension() in EmbeddingModel

### PR DESCRIPTION
**Pull request by Martin7-1**
## Issue

#1080 

## Change

Add `dimension()` method in `EmbeddingModel`. Using an `AbstractEmbeddingModel` to implement `EmbeddingModel` and a common `dimension()` method like below:

```java
    @Override
    public int dimension() {
        if (dimension != null) {
            return dimension;
        }

        return dimensionMap().compute(modelName(), (key, value) -> {
            this.dimension = Optional.ofNullable(value).orElse(embed("test").content().dimension());
            return dimension;
        });
    }
```

Now, existing `{NameOfIntegration}EmbeddingModel` just need to override `dimensionMap()` method to contains known embedding model dimension and `modelName()` method to get embedding model name. 

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)

## Checklist for adding new model integration

- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java [hunk 1]
```diff
@@ -80,10 +80,21 @@ public OpenAiEmbeddingModel(String baseUrl,
...
+    }
+
+    @Override
     public String modelName() {
         return modelName;
     }
 
+    @Override
+    public int dimension() {
+        return dimensions;
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: `dimensions` is an optional parameters, it can be `null`

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    protected abstract String modelName();
+
+    @Override
+    public int dimension() {
+        if (dimension != null) {
+            return dimension;
+        }
+
+        return dimensionMap().compute(modelName(), (key, value) -> {
+            this.dimension = Optional.ofNullable(value).orElse(embed("test").content().dimension());
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]:
```suggestion
            this.dimension = Optional.ofNullable(value).orElseGet(() -> embed("test").content().dimension());
```
Otherwise it will embed every time

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+     */
+    protected abstract String modelName();
+
+    @Override
+    public int dimension() {
+        if (dimension != null) {
+            return dimension;
+        }
+
+        return dimensionMap().compute(modelName(), (key, value) -> {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]:
```suggestion
        return dimensionMap().compute(modelName(), (modelName, dimension) -> {
```
for better clarity

Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: Is the map supposed to be cached?

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: I am having a hard time to understand what this piece of code is doing, could you please rewrite it with simpler if/else conditions?

**Review by ghost — COMMENTED** [review]
@Martin7-1 thank you!

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java (hunk 1 above)
Comment (Martin7-1) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: Now, I remove `dimensions` in `OpenAiEmbeddingModel` because I think `dimension` in `AbstractEmbeddingModel` and `dimenions` in `OpenAiEmbeddingModel` will make user confused. So I delete it and use `dimension` in `AbstractEmbeddingModel` as an alternative way

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java (hunk 3 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: Sure. Sorry for makeing you confused. Already rewrite it!

Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: I'm not sure I know what you mean. This map's purpose is to store known model's name and it's dimension.

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java (hunk 1 above)
Comment (langchain4j) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: No, please don't remove `dimensions` from `OpenAiEmbeddingModel`, it is needed to **_control_** the number of dimensions embedding model returns.

Comment (langchain4j) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: I meant please don't rename `dimensions` into `dimension` in `OpenAiEmbeddingModel`.

Comment (Martin7-1) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: > I meant please don't rename `dimensions` into `dimension` in `OpenAiEmbeddingModel`.

OK. I got it.

Comment (Martin7-1) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: > I meant please don't rename `dimensions` into `dimension` in `OpenAiEmbeddingModel`.

done.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java [hunk 4]
```diff
@@ -44,4 +44,11 @@ default Response<Embedding> embed(TextSegment textSegment) {
      * @return the embeddings.
      */
     Response<List<Embedding>> embedAll(List<TextSegment> textSegments);
+
+    /**
+     * Get embedding model's dimension
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java]:
```suggestion
     * Returns the dimension of the {@link Embedding} produced by this embedding model.
```

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java [hunk 5]
```diff
@@ -44,4 +44,11 @@ default Response<Embedding> embed(TextSegment textSegment) {
...
+     *
+     * @return dimension of the embedding model
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java]:
```suggestion
     * @return dimension of the embedding
```

File: langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/BedrockTitanEmbeddingModel.java [hunk 6]
```diff
@@ -2,24 +2,33 @@
...
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
+import static dev.langchain4j.internal.Utils.getOrDefault;
+
 /**
  * Bedrock Amazon Titan embedding model
  */
-@SuperBuilder
```
Comment (langchain4j) [inline langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/BedrockTitanEmbeddingModel.java]: Please make sure changes in this class are not breaking

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    protected abstract String modelName();
+
+    @Override
+    public int dimension() {
+        if (dimension != null) {
+            return dimension;
+        }
+
+        // get known model's dimension first, otherwise embed "test" to get dimension
+        if (dimensionMap().containsKey(modelName())) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: minor: `dimensionMap()` can be stored in a local variable to avoid creating this map again

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java [hunk 8]
```diff
@@ -44,4 +44,11 @@ default Response<Embedding> embed(TextSegment textSegment) {
...
+     */
+    int dimension();
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java]: This is a breaking change: consider that there are implementations of this interface that are not located in this repository.

What about making this method default?
```java
    default int dimension() {
        return embed("test").content().dimension();
    }
```
In theory this should work for all implementations of this interface.

For specific cases where we know the dimension up-front and want to optimize this method, the implementation class can override this method with the logic similar to what you have implemented in this PR.

File: langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/internal/AbstractBedrockEmbeddingModel.java [hunk 9]
```diff
@@ -25,22 +23,27 @@
 import java.util.stream.Collectors;
 
 import static dev.langchain4j.internal.RetryUtils.withRetry;
+import static dev.langchain4j.internal.Utils.getOrDefault;
 
 /**
  * Abstract bedrock embedding model
  */
-@SuperBuilder
```
Comment (langchain4j) [inline langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/internal/AbstractBedrockEmbeddingModel.java]: same here

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package dev.langchain4j.model.embedding;
+
+import java.util.Map;
+
+/**
+ * An Embedding Model which contains
+ */
+public abstract class AbstractEmbeddingModel implements EmbeddingModel {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]:
```suggestion
public abstract class DimensionAwareEmbeddingModel implements EmbeddingModel {
```
WDYT?

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java [hunk 11]
```diff
@@ -80,6 +80,12 @@ public OpenAiEmbeddingModel(String baseUrl,
         this.tokenizer = getOrDefault(tokenizer, OpenAiTokenizer::new);
     }
 
+    @Override
+    protected Map<String, Integer> dimensionMap() {
+        return OpenAiEmbeddingModelName.embeddingModelDimensionMap();
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: If `dimensions` parameter is specified in the constructor, `OpenAiEmbeddingModel.dimension()` should return the specified value

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+     * embedding model dimension
+     */
+    protected Integer dimension;
+
+    /**
+     * A map contains known model's name and its dimension
+     *
+     * @return A map, key is the common represented model name, value is its dimension
+     */
+    protected abstract Map<String, Integer> dimensionMap();
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: I am not sure there is a need to expose 2 methods (`dimensionMap` and `modelName`).
Can't implementation of this class return dimension value directly, without a need to expose a map?

File: langchain4j-cohere/src/main/java/dev/langchain4j/model/cohere/CohereEmbeddingModel.java [hunk 13]
```diff
@@ -83,4 +86,14 @@ private static TokenUsage getTokenUsage(EmbedResponse response) {
         }
         return null;
     }
+
+    @Override
+    protected Map<String, Integer> dimensionMap() {
+        return new HashMap<>();
```
Comment (langchain4j) [inline langchain4j-cohere/src/main/java/dev/langchain4j/model/cohere/CohereEmbeddingModel.java]: can this be a default implementation in `AbstractEmbeddingModel `? So that there is no need to duplicate this code in many places

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java (hunk 10 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: I think this is great and makes the purpose of this class more clearer. Then maybe not all implmenetations should implement this class, some of them may just implement `EmbeddingModel` (This also avoid breaking changes in `langchain4j-bedrock` module and `langchain4j-workers-ai` module).

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java (hunk 8 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java]: OK. It's a good idea to make it default!

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java (hunk 4 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java]: done.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java (hunk 5 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java]: done.

File: langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/BedrockTitanEmbeddingModel.java (hunk 6 above)
Comment (Martin7-1) [reply langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/BedrockTitanEmbeddingModel.java]: done. Revert it to implement `EmbeddingModel` directly

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java (hunk 7 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: Good idea. Now the `DimensionAwareEmbeddingModel` has `dimensionMap` attribute to cache.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java (hunk 8 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/EmbeddingModel.java]: done.

File: langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/internal/AbstractBedrockEmbeddingModel.java (hunk 9 above)
Comment (Martin7-1) [reply langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/internal/AbstractBedrockEmbeddingModel.java]: done. See above.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java (hunk 10 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: done.

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java (hunk 11 above)
Comment (Martin7-1) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModel.java]: done.

File: langchain4j-cohere/src/main/java/dev/langchain4j/model/cohere/CohereEmbeddingModel.java (hunk 13 above)
Comment (Martin7-1) [reply langchain4j-cohere/src/main/java/dev/langchain4j/model/cohere/CohereEmbeddingModel.java]: done.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java (hunk 12 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: hmmmmm...It's a good question. The difference between `EmbeddingModel` and `DimensionAwareEmbeddingModel` is that the latter cache some known model's dimension and make it more convenient to get embedding's dimension. The purpose to expose `dimensionMap` and `modelName` is just like `EmbeddingStoreWithoutMetadataIT`'s `embeddingModel()` and `embeddingStore()`. I think it makes `dimension()` more standardized and uniform without having to worry about differences between `modelName` and `dimensionMap` in different implementations

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    protected abstract String modelName();
+
+    @Override
+    public int dimension() {
+        if (dimension != null) {
+            return dimension;
+        }
+
+        // get known model's dimension first, otherwise embed "test" to get dimension
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]:
```suggestion
```
Not very helpful comment

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    protected Map<String, Integer> dimensionMap = new HashMap<>();
+
+    /**
+     * A map contains known model's name and its embedding's dimension
+     *
+     * @return A map, key is the common represented model name, value is its embedding's dimension
+     */
+    protected Map<String, Integer> dimensionMap() {
+        return dimensionMap;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]:
```suggestion
        return emptyMap();
```

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+public abstract class DimensionAwareEmbeddingModel implements EmbeddingModel {
+
+    /**
+     * dimension of embedding
+     */
+    protected Integer dimension;
+    /**
+     * dimension map of known embedding model's name and its embedding's dimension
+     */
+    protected Map<String, Integer> dimensionMap = new HashMap<>();
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]:
```suggestion
```
It seems that there is no need to keep/cache this map

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java (hunk 12 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: > The purpose to expose dimensionMap and modelName is just like EmbeddingStoreWithoutMetadataIT's embeddingModel() and embeddingStore(). I think it makes dimension() more standardized and uniform without having to worry about differences between modelName and dimensionMap in different implementations

Not sure I understand, can you please elaborate?

A couple of minor downsides that I see:
- `dimensionMap()` exposes much more information (dimensions for all models) than is required for the task. At some point someone will have a temptation to misuse it.
- having `Map` allows only matching models by exact name, some other implementation-specific operations like "modelName.startsWith("...")" are not possible

Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]: The standard step to get dimension is:

1. check cached dimension, if exist, return it.
2. check known model's dimension, if there is any match, cache and return it.
3. manually embed "test" and cache it.

There are two differences in these steps:

1. different implementations may have different ways of getting model name, e.g. some implementations get it via the modelName attribute, some names may be, as you say, a prefix or something like that.
2. implementations may have different ways of storing known dimensional models, perhaps via enum, perhaps via map, etc., depending on the implementation.

But for `DimensionAwareEmbeddingModel`, all it has to do is to check if there is a corresponding cached dimension based on the modelName, and if there is one, get it directly, if not, embed it manually and get it. So I think their exposure is necessary.

> dimensionMap() exposes much more information (dimensions for all models) than is required for the task. At some point someone will have a temptation to misuse it.

Do you have any good ideas to avoid it? I currently have it set to `protected`, so it is invisible to users. Maybe we should rename it to `getKnownDimension()` to avoid misuse?

> having Map allows only matching models by exact name, some other implementation-specific operations like "modelName.startsWith("...")" are not possible

That's a real downside...Maybe we need a `Function` to calculate the dimension, like `getKnownDimension()`. This allows implementations to implement it more freely.

So the solution may like below:

```java
public int dimension() {
    if (dimension != null) {
        return dimension;
    }
    Integer knownDimension = getKnownDimension();  // different implementations have different ways to get known dimension
    dimension = Optional.ofNullable(knownDimension).orElseGet(() -> embed("test").content().dimension());
    return dimension;
}

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java (hunk 14 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]: done.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java (hunk 12 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java]:
```java
public int dimension() {
    if (dimension != null) {
        return dimension;
    }
    Integer knownDimension = getKnownDimension();  // different implementations have different ways to get known dimension
    dimension = Optional.ofNullable(knownDimension).orElseGet(() -> embed("test").content().dimension());
    return dimension;
}
```

Maybe this will make implementations to get known dimension more freely

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModelName.java [hunk 17]
```diff
@@ -1,5 +1,8 @@
 package dev.langchain4j.model.openai;
 
+import java.util.HashMap;
+import java.util.Map;
+
 public enum OpenAiEmbeddingModelName {
 
     TEXT_EMBEDDING_3_SMALL("text-embedding-3-small"),
```
Comment (langchain4j) [inline langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModelName.java]:
```suggestion
    TEXT_EMBEDDING_3_SMALL("text-embedding-3-small", 1536),
```
WDYT about putting it here instead of map?

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/client/AbstractWorkersAIModel.java [hunk 18]
```diff
@@ -30,22 +30,19 @@ public abstract class AbstractWorkersAIModel {
...
-     * @param modelName
-     *      model name.
-     * @param apiToken
-     *      api apiToken from .
+     * @param accountId account identifier.
+     * @param modelName model name.
+     * @param apiToken  api apiToken from .
      */
-    public AbstractWorkersAIModel(String accountId, String modelName, String apiToken) {
+    protected AbstractWorkersAIModel(String accountId, String modelName, String apiToken) {
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/client/AbstractWorkersAIModel.java]: why this change?

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiEmbeddingModelName.java [hunk 19]
```diff
@@ -1,5 +1,8 @@
 package dev.langchain4j.model.azure;
 
+import java.util.HashMap;
+import java.util.Map;
+
 public enum AzureOpenAiEmbeddingModelName {
 
     TEXT_EMBEDDING_3_SMALL("text-embedding-3-small", "text-embedding-3-small"), // alias for the latest text-embedding-3-small model
```
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiEmbeddingModelName.java]:
```suggestion
    TEXT_EMBEDDING_3_SMALL("text-embedding-3-small", "text-embedding-3-small", 1536), // alias for the latest text-embedding-3-small model
```
WDYT about putting it here instead of map?

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+     * dimension of embedding
+     */
+    protected Integer dimension;
+
+    /**
+     * Returns the dimension of known {@link Embedding} produced by this embedding model. if it's unknown, return null
+     *
+     * @return the dimension of known {@link Embedding}, null if unknown.
+     */
+    protected Integer getKnownDimension() {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]:
```suggestion
    protected Integer knownDimension() {
```

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+ */
+public abstract class DimensionAwareEmbeddingModel implements EmbeddingModel {
+
+    /**
+     * dimension of embedding
+     */
+    protected Integer dimension;
+
+    /**
+     * Returns the dimension of known {@link Embedding} produced by this embedding model. if it's unknown, return null
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]:
```suggestion
     * When known (e.g., can be derived from the model name), returns the dimension of the {@link Embedding} produced by this embedding model. Otherwise, it returns {@code null}.
```

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+     *
+     * @return the dimension of known {@link Embedding}, null if unknown.
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]:
```suggestion
     * @return the known dimension of the {@link Embedding}, or {@code null} if unknown.
```

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiEmbeddingModel.java
Comment (langchain4j) [inline langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiEmbeddingModel.java:1]: You've removed gecko mappings in the previous commit, probably by accident?

**Review by ghost — COMMENTED** [review]
@Martin7-1 I like this new solution much more, thank you!

Just a few minor things and we can merge!

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModelName.java (hunk 17 above)
Comment (Martin7-1) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModelName.java]: Good idea. But there is one thing to consider:

In `OpenAiEmbeddingModel`, String is used to represent modelName instead of OpenAiEmbeddingModelName, so if we want to add dimension to an enum, maybe we need a map in `OpenAiEmbeddingModelName` like below:

```java
    private static final Map<String, Integer> KNOWN_DIMENSION = new HashMap<>(OpenAiEmbeddingModelName.values().length);
    
    static {
        for (OpenAiEmbeddingModelName embeddingModelName : OpenAiEmbeddingModelName.values()) {
            KNOWN_DIMENSION.put(embeddingModelName.toString(), embeddingModelName.dimension());
        }
    }

    public static Integer knownDimension(String modelName) {
        return KNOWN_DIMENSION.get(modelName);
    }
```

WDYT?

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiEmbeddingModel.java
Comment (Martin7-1) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiEmbeddingModel.java:1]: Sorry for that. Now I implement it like `OpenAiEmbeddingModelName`

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java (hunk 22 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]: done.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java (hunk 21 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]: done.

File: langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java (hunk 20 above)
Comment (Martin7-1) [reply langchain4j-core/src/main/java/dev/langchain4j/model/embedding/DimensionAwareEmbeddingModel.java]: done.

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/client/AbstractWorkersAIModel.java (hunk 18 above)
Comment (Martin7-1) [reply langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/client/AbstractWorkersAIModel.java]: revert. Maybe it's problem of IntelliJ.

**Comment by Martin7-1** [comment]
The CI failed maybe because the `main` branch's commit. @langchain4j

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiEmbeddingModel.java
Comment (langchain4j) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiEmbeddingModel.java:1]: Thanks!

File: langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModelName.java (hunk 17 above)
Comment (langchain4j) [reply langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/OpenAiEmbeddingModelName.java]: Sounds good!

**Review by ghost — APPROVED** [review]
@Martin7-1 good job, thank you!

**Comment by langchain4j** [comment]
@Martin7-1 could you please also update embedding models in https://github.com/langchain4j/langchain4j-embeddings ?

**Comment by Martin7-1** [comment]
> @Martin7-1 could you please also update embedding models in https://github.com/langchain4j/langchain4j-embeddings ?

Sure!
